x-pack/plugins/security_solution/server/endpoint/routes/actions/file_upload_handler.ts

```diff
@@ -0,0 +1,136 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { RequestHandler } from '@kbn/core/server';
+import { createFile, deleteFile, setFileActionId } from '../../services';
+import type {
+  ResponseActionUploadParameters,
+  ResponseActionUploadOutputContent,
+} from '../../../../common/endpoint/types';
+import { UPLOAD_ROUTE } from '../../../../common/endpoint/constants';
+import {
+  type UploadActionRequestBody,
+  UploadActionRequestSchema,
+} from '../../../../common/endpoint/schema/actions';
+import { withEndpointAuthz } from '../with_endpoint_authz';
+import type {
+  SecuritySolutionPluginRouter,
+  SecuritySolutionRequestHandlerContext,
+  HapiReadableStream,
+} from '../../../types';
+import type { EndpointAppContext } from '../../types';
+import { errorHandler } from '../error_handler';
+
+export const registerActionFileUploadRoute = (
+  router: SecuritySolutionPluginRouter,
+  endpointContext: EndpointAppContext
+) => {
+  if (!endpointContext.experimentalFeatures.responseActionUploadEnabled) {
+    return;
+  }
+
+  const logger = endpointContext.logFactory.get('uploadAction');
+
+  router.post(
+    {
+      path: UPLOAD_ROUTE,
+      validate: UploadActionRequestSchema,
+      options: {
+        authRequired: true,
+        tags: ['access:securitySolution'],
+        body: {
+          accepts: ['multipart/form-data'],
+          output: 'stream',
+          maxBytes: endpointContext.serverConfig.maxUploadResponseActionFileBytes,
+        },
+      },
+    },
+    withEndpointAuthz(
+      { all: ['canWriteFileOperations'] },
+      logger,
+      getActionFileUploadHandler(endpointContext)
+    )
+  );
+};
+
+export const getActionFileUploadHandler = (
+  endpointContext: EndpointAppContext
+): RequestHandler<never, never, UploadActionRequestBody, SecuritySolutionRequestHandlerContext> => {
+  const logger = endpointContext.logFactory.get('uploadAction');
+  const maxFileBytes = endpointContext.serverConfig.maxUploadResponseActionFileBytes;
+
+  return async (context, req, res) => {
+    const user = endpointContext.service.security?.authc.getCurrentUser(req);
+    const esClient = (await context.core).elasticsearch.client.asInternalUser;
+    const fileStream = req.body.file as HapiReadableStream;
+    const { file: _, parameters: userParams, ...actionPayload } = req.body;
+    const uploadParameters: ResponseActionUploadParameters = {
+      ...userParams,
+      file: {
+        file_id: '',
+        file_name: '',
+        sha256: '',
+        size: 0,
+      },
+    };
+
+    try {
+      const createdFile = await createFile({
+        esClient,
+        logger,
+        fileStream,
+        agents: actionPayload.endpoint_ids,
+        maxFileBytes,
+      });
+
+      uploadParameters.file.file_id = createdFile.file.id;
+      uploadParameters.file.file_name = createdFile.file.name;
+      uploadParameters.file.sha256 = createdFile.file.hash?.sha256;
```

Comment:
```suggestion
      uploadParameters.file.sha256 = createdFile.file.hash.sha256;
```

Reply:
`file.hash` is an optional attribute.

Reply:
Oh I see. I saw it typed as a required `string` in the PR so I suggested the change.